### PR TITLE
[LW] Wire cache into transactions

### PR DESCRIPTION
**Pull request by gmaretic**
**Goals (and why)**:

**Implementation Description (bullets)**:

**Testing (What was existing testing like?  What have you done to improve it?)**:

**Concerns (what feedback would you like?)**:

**Where should we start reviewing?**:

**Priority (whenever / two weeks / yesterday)**:

### CODE REVIEW

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 1]
```diff
@@ -1569,12 +1581,14 @@ protected boolean shouldDeleteAndRollback() {
     @Override
     public final void delete(TableReference tableRef, Set<Cell> cells) {
         putInternal(tableRef, Cells.constantValueMap(cells, PtBytes.EMPTY_BYTE_ARRAY));
+        cells.forEach(cell -> cache.write(tableRef, cell, CacheValue.empty()));
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1584]: I can add a method for `cache.delete`, would make this nicer.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 2]
```diff
@@ -1569,12 +1581,14 @@ protected boolean shouldDeleteAndRollback() {
...
     }
 
     @Override
     public void put(TableReference tableRef, Map<Cell, byte[]> values) {
         ensureNoEmptyValues(values);
         putInternal(tableRef, values);
+        values.forEach((cell, value) -> cache.write(tableRef, cell, CacheValue.of(value)));
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1591]: didn't realise this was done in a map, I can make this just accept a map instead of individual values.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 3]
```diff
@@ -1814,6 +1828,8 @@ private void commitWrites(TransactionService transactionService) {
                         "commitPutCommitTs",
                         () -> putCommitTimestamp(commitTimestamp, commitLocksToken, transactionService));
 
+                lockWatchValueCache.updateCacheOnCommit(cache.getDigest(), getStartTimestamp());
+
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1832]: This might live better in the same place that `LockWatchEventCache` does its update on commit, because then we can batch (this is the central cache you're updating, which means this could slow things done if we're not careful).

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java [hunk 4]
```diff
@@ -184,6 +192,10 @@ protected boolean shouldStopRetrying(int numTimesFailed) {
                 timelockService.startIdentifiedAtlasDbTransactionBatch(conditions.size());
         Preconditions.checkState(conditions.size() == responses.size(), "Different number of responses and conditions");
         try {
+            lockWatchValueCache.processStartTransactions(responses.stream()
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java:195]: again, this should probably live in the same place as `LockWatchEventCache` and its processing.

Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java:195]: there may even be an argument of pairing the two of them in some handy interface.

**Review by Jolyon-S — COMMENTED** [review]
There may be some improvements we can make, comments below.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java (hunk 1 above)
Comment (gmaretic) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1584]: Yup, that would make sense 👍

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java (hunk 2 above)
Comment (gmaretic) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1591]: 👍

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java (hunk 4 above)
Comment (gmaretic) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java:195]: Yeah, there is a comment, this should live in transaction starter, but I need to do more annoying wiring first

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java (hunk 3 above)
Comment (gmaretic) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1832]: You mean in the CommitTimestampGetters? That probably makes sense, yeah. Just to double check -- we only actually update with the reads, so even if the transaction fails after we do this update, it's still correct, right?

Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:1832]: the former: yes (I think that's what it is called these days).

the latter: only update reads, and that _should_ happen after we verify the PCC, so if there were R-W conflicts, those reads should not end up in the cache either (we'll have to be careful about which part of the lifecycle that happens in).

It is worth noting that, internally, we actually only update the cache once the transaction has fully committed successfully, but I think it is technically fine to do it during commit here (with the caveat above that the PCC is done at the correct time).

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueCache.java [hunk 5]
```diff
@@ -31,10 +32,16 @@
  * the central cache (taking in to account the since-locked descriptors), as well as checking for conflicts for
  * serializable transactions by adding a check in the {@link com.palantir.atlasdb.transaction.api.PreCommitCondition}.
  */
-public interface LockWatchValueCache {
+public interface LockWatchValueCache extends ValueCacheUpdater {
+    @Override
```
Comment (gmaretic) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueCache.java:36]: I hate the atlas pattern of hiding inherited methods

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java (hunk 4 above)
Comment (gmaretic) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java:194]: Did the wiring, will look into unifying the caches to make code slightly nicer

File: atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java [hunk 6]
```diff
@@ -439,77 +438,6 @@ public void testLockAfterGet() throws Exception {
         mockery.assertIsSatisfied();
     }
 
-    @Ignore("Was ignored long ago, and now we need to fix the mocking logic.")
```
Comment (gmaretic) [inline atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java:442]: drive by

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueCache.java [hunk 7]
```diff
@@ -31,10 +32,16 @@
...
     void processStartTransactions(Set<Long> startTimestamps);
 
-    void updateCacheOnCommit(ValueDigest digest, long startTs);
+    @Override
+    void updateCacheOnCommit(Set<Long> startTimestamps);
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueCache.java:40]: You can merge in https://github.com/palantir/atlasdb/pull/5408/files for this change (ideally, we get that merged before this).

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueCache.java [hunk 8]
```diff
@@ -31,10 +32,16 @@
...
+
+    default void processFailedTransactions(Set<Long> startTimestamp) {
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueCache.java:42]: hmm, I suppose we need to remove the state if the transaction failed? I can implement this in a subsequent PR as you mentioned.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransaction.java [hunk 9]
```diff
@@ -304,15 +305,7 @@ private void updateRangeEnd(Cell end) {
...
-        try {
-            return getWithLoader(
-                            tableRef,
-                            cells,
-                            (tableReference, toRead) -> Futures.immediateFuture(super.get(tableRef, toRead)))
-                    .get();
-        } catch (InterruptedException | ExecutionException e) {
-            throw Throwables.rewrapAndThrowUncheckedException(e.getCause());
-        }
+        return super.get(tableRef, cells);
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SerializableTransaction.java:308]: Why aren't we getting with loader? We'll surely still need to do R-W conflicts for normal remote reads?

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ShouldNotDeleteAndRollbackTransaction.java [hunk 10]
```diff
@@ -85,6 +87,7 @@ public ShouldNotDeleteAndRollbackTransaction(
                 keyValueService,
                 null,
                 null,
+                new LockWatchValueCacheImpl(NoOpLockWatchEventCache.create()),
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/ShouldNotDeleteAndRollbackTransaction.java:90]: I think we should create a no-op value cache.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java [hunk 11]
```diff
@@ -771,13 +778,17 @@ private void extractLocalWritesForRow(
...
+                cells,
+                (table, uncached) ->
+                        getInternal("get", tableRef, cells, immediateKeyValueService, immediateTransactionService));
     }
 
     @Override
     @Idempotent
     public ListenableFuture<Map<Cell, byte[]>> getAsync(TableReference tableRef, Set<Cell> cells) {
+        // todo(gmaretic): make this use cache as well?
         return getInternal("getAsync", tableRef, cells, keyValueService, defaultTransactionService);
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:792]: the cache's `get` method is literally just a `AtlasFutures.getUnchecked` on the `getInternal` - not against just renaming and exposing that so that we have async as well.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java [hunk 12]
```diff
@@ -321,6 +333,8 @@ protected Transaction createTransaction(
                 keyValueService,
                 timelockService,
                 lockWatchManager,
+                // cannot use caching if we do not talk to timelock
+                new LockWatchValueCacheImpl(NoOpLockWatchEventCache.create()),
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java:337]: Again, a no-op cache would be nice.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueCacheImplTest.java [hunk 13]
```diff
@@ -78,7 +78,7 @@
             ImmutableSet.of(LockWatchReferences.entireTable(TABLE.getQualifiedName())));
 
     private LockWatchEventCache eventCache;
-    private LockWatchValueCache valueCache;
+    private LockWatchValueCacheImpl valueCache;
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/cache/LockWatchValueCacheImplTest.java:81]: Why is this necessary?

File: lock-api/src/main/java/com/palantir/lock/cache/ValueCacheUpdater.java [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.cache;
+
+import java.util.Set;
+
+public interface ValueCacheUpdater {
```
Comment (Jolyon-S) [inline lock-api/src/main/java/com/palantir/lock/cache/ValueCacheUpdater.java:21]: annoying that we need this, but I buy it.

Comment (Jolyon-S) [reply lock-api/src/main/java/com/palantir/lock/cache/ValueCacheUpdater.java:21]: Although actually there might be a decomp where we pair the caches _and_ don't need this.

File: lock-api/src/main/java/com/palantir/lock/cache/ValueCacheUpdater.java [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    ValueCacheUpdater NO_OP = new ValueCacheUpdater() {
```
Comment (Jolyon-S) [inline lock-api/src/main/java/com/palantir/lock/cache/ValueCacheUpdater.java:22]: Ugly, but I think this will do _for now_. Can we add a TODO to maybe revisit this?

File: lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java [hunk 16]
```diff
@@ -62,7 +63,8 @@ public static RemoteTimelockServiceAdapter create(
                 rpcClient,
                 conjureClient,
                 new LegacyLeaderTimeGetter(conjureClient),
-                RequestBatchersFactory.create(lockWatchEventCache, namespace, Optional.empty()));
+                RequestBatchersFactory.create(
+                        lockWatchEventCache, ValueCacheUpdater.NO_OP, namespace, Optional.empty()));
```
Comment (Jolyon-S) [inline lock-api/src/main/java/com/palantir/lock/client/RemoteTimelockServiceAdapter.java:67]: Checked that everywhere that calls this has a no-op event cache. In that case we may as well surely just inline the no op event cache here too?

File: lock-api/src/test/java/com/palantir/lock/client/TransactionStarterTest.java [hunk 17]
```diff
@@ -55,6 +57,7 @@
 
     private final StartTransactionsLockWatchEventCache lockWatchEventCache =
             spy(StartTransactionsLockWatchEventCache.createForTests());
+    private final ValueCacheUpdater valueCache = mock(ValueCacheUpdater.class);
```
Comment (Jolyon-S) [inline lock-api/src/test/java/com/palantir/lock/client/TransactionStarterTest.java:60]: we do have a handy no-op implementation we could use, but this is fine, don't really care.

**Review by Jolyon-S — CHANGES_REQUESTED** [review]
I think we could technically merge like this or something similar, but I wonder if that'd actually be more overhead for us than upside: we'd have to revisit and re-wire things. I honestly think that maybe we should just couple the caches now instead of later (happy to do that in a PR if you want).

File: lock-api/src/main/java/com/palantir/lock/cache/ValueCacheUpdater.java (hunk 15 above)
Comment (Jolyon-S) [reply lock-api/src/main/java/com/palantir/lock/cache/ValueCacheUpdater.java:22]: I've made this better in https://github.com/palantir/atlasdb/pull/5409

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java (hunk 11 above)
Comment (gmaretic) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:788]: That's probably outiside the scope here

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/NoOpLockWatchManager.java [hunk 18]
```diff
@@ -16,23 +16,28 @@
...
 import com.palantir.lock.watch.LockWatchVersion;
 import com.palantir.lock.watch.TransactionsLockWatchUpdate;
 import java.util.Optional;
 import java.util.Set;
 
-public final class NoOpLockWatchManager extends LockWatchManager {
-    private final LockWatchEventCache eventCache;
+public final class NoOpLockWatchManager extends LockWatchManagerInternal {
+    private final LockWatchCache cache;
+    private final TransactionScopedCache transactionScopedCache;
```
Comment (Jolyon-S) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/NoOpLockWatchManager.java:32]: I expect I'll find out why this is happening - the no op value cache should happily hand out no op transaction caches, so this feels weird.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManagerImpl.java [hunk 19]
```diff
@@ -36,40 +43,54 @@
...
-public final class LockWatchManagerImpl extends LockWatchManager implements AutoCloseable {
+public final class LockWatchManagerImpl extends LockWatchManagerInternal {
 
     private static final Logger log = LoggerFactory.getLogger(LockWatchManagerImpl.class);
 
     private final Set<LockWatchReferences.LockWatchReference> referencesFromSchema;
     private final Set<LockWatchReferences.LockWatchReference> lockWatchReferences = ConcurrentHashMap.newKeySet();
-    private final LockWatchEventCache lockWatchEventCache;
+    private final LockWatchCache lockWatchCache;
+    private final LockWatchValueScopingCache valueScopingCache;
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManagerImpl.java:53]: There should be a method on LWC to get LWVSC, so no need to keep as a field.

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java (hunk 11 above)
Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java:788]: https://github.com/palantir/atlasdb/pull/5413 implements `getAsync`, but we can merge that in the next PR (probably too painful to do _yet another_ rebase here :P)

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java [hunk 20]
```diff
@@ -240,7 +237,7 @@ public Transaction delegate() {
             try {
                 result = runTaskThrowOnConflict(wrappedTask, tx);
             } finally {
-                lockWatchEventCache.removeTransactionStateFromCache(getTimestamp());
+                lockWatchManager.removeTransactionStateFromCache(getTimestamp());
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java:240]: It's worth noting a discrepancy in the implementation that we might need to be careful of: LWEC requires this to be called to clear state, whereas LWVC will clear the state on commit (and so this is only for use in an exceptional context, although it doesn't matter if it is called also). Just need to make sure that we definitely do the commit method before this, although I expect we probably do.

File: atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManagerImplTest.java [hunk 21]
```diff
@@ -40,6 +41,9 @@
     @Mock
     private LockWatchEventCache lockWatchEventCache;
 
+    @Mock
+    private LockWatchValueScopingCache valueScopingCache;
```
Comment (Jolyon-S) [inline atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManagerImplTest.java:45]: Probably doing a couple of really basic tests confirming that the methods pass through to the right caches

File: lock-api-objects/src/main/java/com/palantir/lock/watch/LockWatchStateUpdate.java [hunk 22]
```diff
@@ -33,6 +34,7 @@
     @JsonSubTypes.Type(value = LockWatchStateUpdate.Success.class, name = LockWatchStateUpdate.Success.TYPE),
     @JsonSubTypes.Type(value = LockWatchStateUpdate.Snapshot.class, name = LockWatchStateUpdate.Snapshot.TYPE)
 })
+@DoNotMock
```
Comment (Jolyon-S) [inline lock-api-objects/src/main/java/com/palantir/lock/watch/LockWatchStateUpdate.java:37]: I've never seen this before - why do we need it?

File: lock-api/src/main/java/com/palantir/lock/watch/LockWatchCacheImpl.java [hunk 23]
```diff
@@ -34,6 +29,10 @@ public LockWatchCacheImpl(LockWatchEventCache eventCache, LockWatchValueCache va
         this.valueCache = valueCache;
     }
 
+    public static LockWatchCache noop() {
```
Comment (Jolyon-S) [inline lock-api/src/main/java/com/palantir/lock/watch/LockWatchCacheImpl.java:32]: Since we generally call classes `NoOp`,  that implies this should be `noOp()`

**Review by Jolyon-S — COMMENTED** [review]
Nearly there, just a few small style things.

I think the reason I'm being so pedantic on this PR is that this actually just wires production code to use the value cache, which I'm probably being more scared than I need to be when it comes to enabling!

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/NoOpLockWatchManager.java (hunk 18 above)
Comment (Jolyon-S) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/NoOpLockWatchManager.java:32]: (Update: didn't work out why you did this - why not just create no op caches on the fly?)

Comment (gmaretic) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/NoOpLockWatchManager.java:32]: We can inline, sure

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManagerImpl.java (hunk 19 above)
Comment (gmaretic) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManagerImpl.java:53]: There cannot be one, for the same reason that LWVSC exists in the first place

Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManagerImpl.java:53]: ah damn

Comment (Jolyon-S) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchManagerImpl.java:53]: (Might be worth noting that or commenting that we do _not_ need to update the transaction state in both LWC and LWVSC).

File: atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java (hunk 20 above)
Comment (gmaretic) [reply atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java:240]: The only time we do not do the commit method before is if the transaction fails, and we will never do it

File: lock-api-objects/src/main/java/com/palantir/lock/watch/LockWatchStateUpdate.java (hunk 22 above)
Comment (Jolyon-S) [reply lock-api-objects/src/main/java/com/palantir/lock/watch/LockWatchStateUpdate.java:37]: Discussed - fine, I buy it.

Comment (gmaretic) [reply lock-api-objects/src/main/java/com/palantir/lock/watch/LockWatchStateUpdate.java:37]: Because mocking this makes tests explode since even the noop implementations depend on being able to read stuff

File: lock-api/src/main/java/com/palantir/lock/watch/LockWatchCacheImpl.java (hunk 23 above)
Comment (gmaretic) [reply lock-api/src/main/java/com/palantir/lock/watch/LockWatchCacheImpl.java:32]: I don't even know why I do these things when I know you will complain about it 🗡️

Comment (Jolyon-S) [reply lock-api/src/main/java/com/palantir/lock/watch/LockWatchCacheImpl.java:32]: 🔥

**Review by Jolyon-S — APPROVED** [review]
As far as I can tell, for the average user, _nothing_ should change here:
1. LWEC works as before
2. Reads will now go through the cache - there is overhead in creating one, but then it will just read immediately because no tables are watched.
